geode-core/src/test/java/org/apache/geode/internal/cache/CacheEventListenerDUnitTest.java

```diff
@@ -0,0 +1,541 @@
+    // The third createEntry causes 2 update events, etc (in the first 2 VMs)
+    Host host = Host.getHost(0);
+    int vmCount = host.getVMCount();
+    for (int i = 0; i < vmCount; i++) {
+      host.getVM(i).invoke(() -> createEntry(name, listener));
+
+      expectedCreateInv++;
+      expectedUpdateInv += (i + 1);
+
+      logger.info("TESTOBJECTADDEDREPLACED: verifying callbacks for VM " + i);
+      logger.info("TESTOBJECTADDEDREPLACED: done verifying callbacks for VM " + i);
+    }
+
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_1))
+        .as("Wrong number of 'afterCreate' listener invocations").isEqualTo(expectedCreateInv);
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_2))
+        .as("Wrong number of 'afterUpdate' listener invocations").isEqualTo(expectedUpdateInv);
+
+    assertThat(newValue).isEqualTo(0);
+    assertThat(oldValue).isEqualTo(0);
+
+    // test replace events
+    int replaceValue = 1;
+    replaceEntry(name, replaceValue);
+    expectedUpdateInv += numVMs;
+
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_1))
+        .as("Wrong number of 'afterCreate' listener invocations").isEqualTo(expectedCreateInv);
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_2))
+        .as("Wrong number of 'afterUpdate' listener invocations").isEqualTo(expectedUpdateInv);
+
+    assertThat(getOldValue()).isEqualTo(replaceValue - 1);
+    assertThat(getNewValue()).isEqualTo(replaceValue);
+
+    for (int i = 0; i < vmCount; i++) {
+      Object vmOldValue = host.getVM(i).invoke(() -> getOldValue());
+      Object vmNewValue = host.getVM(i).invoke(() -> getNewValue());
+      assertThat(vmOldValue).as("oldValue is wrong in vm " + i).isEqualTo(replaceValue - 1);
+      assertThat(vmNewValue).as("newValue is wrong in vm " + i).isEqualTo(replaceValue);
+    }
+
+    replaceValue++;
+    int finalReplaceValue = replaceValue;
+    host.getVM(0).invoke(() -> this.replaceEntry(name, finalReplaceValue));
+    expectedUpdateInv += numVMs;
+
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_1))
+        .as("Wrong number of 'afterCreate' listener invocations").isEqualTo(expectedCreateInv);
+    assertThat(sharedCountersRule.getTotal(NUM_INVOCATIONS_2))
+        .as("Wrong number of 'afterUpdate' listener invocations").isEqualTo(expectedUpdateInv);
+
+    assertThat(getOldValue()).isEqualTo(replaceValue - 1);
+    assertThat(getNewValue()).isEqualTo(replaceValue);
+
+    for (int i = 0; i < vmCount; i++) {
+      Object vmOldValue = host.getVM(i).invoke(() -> getOldValue());
+      Object vmNewValue = host.getVM(i).invoke(() -> getNewValue());
+      assertThat(vmOldValue).as("oldValue is wrong in vm " + i).isEqualTo(replaceValue - 1);
+      assertThat(vmNewValue).as("newValue is wrong in vm " + i).isEqualTo(replaceValue);
+    }
+  }
+
+  @Test
+  public void testObjectInvalidated() throws Exception {
+    int timeToLive = 5; // seconds
```

Comment:
Why 5 seconds? At the most it seems like it should be 1 second. I think we have some product support to change these units to milliseconds so that these tests could run faster.

Reply:
I didn't write the test. I'm updating it to JUnit 4 and adding it to Geode.